test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs

```diff
@@ -1881,6 +1881,177 @@ public async Task TransitiveDependenciesFromNonRootLibraries_AreIgnored()
+            foreach (var node in nonCentralTransitiveNodes)
+            {
+                Assert.Equal(0, node.ParentNodes.Count);
+                Assert.True(staticEmptyList == node.ParentNodes); //All nonCentralTransitiveNodes have ParentNodes pointing to the same empty list.
+            }
+
+            //check if InnerNodes of nodesWithEmptyInnerNodes are pointing to the static empty list correctly
+            var nodesWithEmptyInnerNodes = allNodes.Where(n => n.InnerNodes.Count == 0).ToList();
+            Assert.Equal(3, nodesWithEmptyInnerNodes.Count);
+            Assert.True(nodesWithEmptyInnerNodes.Any(n => n.Item.Key.Name == "C"));
+            Assert.True(nodesWithEmptyInnerNodes.Any(n => n.Item.Key.Name == "D"));
+            Assert.True(nodesWithEmptyInnerNodes.Any(n => n.Item.Key.Name == "G"));
+            Assert.True(nodesWithEmptyInnerNodes.Where(n => n.Item.Key.Name == "G").Single().InnerNodes == staticEmptyList);
+
+            Assert.Equal(1, result.Downgrades.Count);
+        }
+
+        /// <summary>
+        ///   -> B 1.0.0 -> C 1.0.0 
+        /// A -> C 2.0.0
+        ///   -> D 1.0.0 -> E 1.0.0
+        /// </summary>
+        [Fact]
+        public async Task WalkAsync_WithNoCPM_InnerNodesAndParentNodesCreatedCorrectly()
+        {
+            var framework = NuGetFramework.Parse("net45");
+
+            var context = new TestRemoteWalkContext();
+            var provider = new DependencyProvider();
+            var version100 = "1.0.0";
+            var version200 = "2.0.0";
+
+            // A -> B 1.0.0 -> C 1.0.0
+            provider.Package("A", version100)
+                    .DependsOn("B", version100);
+            provider.Package("B", version100)
+                   .DependsOn("C", version100);
+
+            // A -> C 2.0.0
+            provider.Package("A", version100)
+                    .DependsOn("C", version200);
+
+            // A -> D 1.0.0 -> E 1.0.0
+            provider.Package("A", version100)
+                    .DependsOn("D", version100);
+            provider.Package("D", version100)
+                    .DependsOn("E", version100);
+
+            context.LocalLibraryProviders.Add(provider);
+            var walker = new RemoteDependencyWalker(context);
+
+            // Act
+            var rootNode = await DoWalkAsync(walker, "A", framework);
+
+            // Assert
+            AnalyzeResult<RemoteResolveResult> result = rootNode.Analyze();
+            var allNodes = GetAllNodes(rootNode);
+            Assert.Equal(5, allNodes.Count);
+            Assert.Equal(3, rootNode.InnerNodes.Count);
+
+            //check if ParentNodes of allnodes are pointing to the static empty list correctly
+            var staticEmptyList = allNodes.First().ParentNodes;
+            foreach (var node in allNodes)
+            {
+                Assert.Equal(0, node.ParentNodes.Count);
+                Assert.True(staticEmptyList == node.ParentNodes);
```

Comment:
Would this better if we directly compare to `staticEmptyList` instance in GraphNode.cs?

Reply:
Good suggestion! Updated accordingly.